update.sh

```diff
@@ -45,21 +54,38 @@ chmod +x rhino-deinst
 sudo mv rhino-deinst /usr/bin
 
 # Automatically install the latest Linux kernel onto the system if it has not been installed already.
-if [[ ! -f "$HOME/.rhino/config/5-18-11" ]]; then
-    cd ~/rhinoupdate/kernel/
-    wget -q --show-progress --progress=bar:force https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.18.11/amd64/CHECKSUMS &
-    wget -q --show-progress --progress=bar:force https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.18.11/amd64/linux-headers-5.18.11-051811-generic_5.18.11-051811.202207121541_amd64.deb &
-    wget -q --show-progress --progress=bar:force https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.18.11/amd64/linux-headers-5.18.11-051811_5.18.11-051811.202207121541_all.deb &
-    wget -q --show-progress --progress=bar:force https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.18.11/amd64/linux-image-unsigned-5.18.11-051811-generic_5.18.11-051811.202207121541_amd64.deb &
-    wget -q --show-progress --progress=bar:force https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.18.11/amd64/linux-modules-5.18.11-051811-generic_5.18.11-051811.202207121541_amd64.deb &
+if [[ ! -f "$HOME/.rhino/config/$(echo $KERNEL_VERSION | sed s/'\.'/'-'/g)" ]]; then
+    cd "$CACHE_PATH"
+    checksum_path="$HOME/.rhino/config/${KERNEL_VERSION}-CHECKSUMS"
+
+    # Always fetch the checksums
+    wget -q --show-progress --progress=bar:force -O "$checksum_path" https://kernel.ubuntu.com/~kernel-ppa/mainline/v$KERNEL_VERSION/amd64/CHECKSUMS &
+
+    # Fetch only missing files
+    for kdeb in $KERNEL_DEBS ; do
+      if ! [ -f $kdeb ] ; then
+        url="https://kernel.ubuntu.com/~kernel-ppa/mainline/v$KERNEL_VERSION/amd64/$kdeb"
+        sudo wget -q --show-progress --progress=bar:force -O "partial/$kdeb" "$url" &
+      fi
+    done
     wait
-    
+
+    # Move fetched files and remember fully qualified paths
+    full_kdebs=""
+    for kdeb in $KERNEL_DEBS ; do
+      full_kdebs="$full_kdebs $CACHE_PATH/$kdeb"
+      if ( sudo test -f "partial/$kdeb" ) ; then
```

Comment:
Why are you doing this in a subshell and why are you using sudo?
```suggestion
      if [[ -f "partial/$kdeb" ]]; then
```

Reply:
The `partial` directory is not readable by an ordinary user (at least that is what I saw in testing) so this needs to be wrapped with `sudo` instead of a plain test. I'll add a comment for this.

Once you get back to me on this one I'll update and resolve the other bits of feedback.

Reply:
@Henryws Did you need anything further here?

Reply:
Oh my god, sorry for not responding..... ok sure, you can keep the subshell